### PR TITLE
[ES|QL] Refactor data type conversions between String/UnsignedLong and other data types

### PR DESCRIPTION
**Pull request by fang-xing-esql**
This is a subtask of #104596.

Two items are covered in the PR:

Refactor data type conversions between String and other data types.
Refactor data type conversions between UnsignedLong and other data types.

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToLong.java [hunk 1]
```diff
@@ -81,16 +82,16 @@ static long fromBoolean(boolean bool) {
...
-            return Long.parseLong(asString);
-        } catch (NumberFormatException nfe) {
+            return stringToLong(asString);
+        } catch (InvalidArgumentException iae) {
             try {
-                return fromDouble(Double.parseDouble(asString));
+                return fromDouble(stringToDouble(asString));
             } catch (Exception e) {
-                throw nfe;
+                throw iae;
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToLong.java:94]: Potential optimization is to create a dedicated method that does `stringToSmallestInteger` and look into custom parsing that avoids having to catch exceptions before falling back. Can be a separate PR but in that case please raise an issue for it.

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Log10Tests.java [hunk 2]
```diff
@@ -54,7 +56,7 @@ public static Iterable<Object[]> parameters() {
             suppliers,
             "Log10UnsignedLongEvaluator[val=" + read + "]",
             DataTypes.DOUBLE,
-            ul -> Math.log10(ul == null ? null : NumericUtils.unsignedLongToDouble(NumericUtils.asLongUnsigned(ul))),
+            ul -> Math.log10(ul == null ? null : unsignedLongToDouble(NumericUtils.asLongUnsigned(ul))),
```
Comment (bpintea) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Log10Tests.java:59]: Would we want here an `unsignedLongToDouble(long)` util?

**Review by bpintea — APPROVED** [review]
Lgtm, left some notes, more relevant being around structuring of the punctual conversion utils.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java [hunk 3]
```diff
@@ -185,6 +195,45 @@ public static long chronoToLong(long dateTime, ChronoField chronoField, ZoneId z
         return Instant.ofEpochMilli(dateTime).atZone(zone).getLong(chronoField);
     }
 
+    /**
+     * The following conversions are between String and other data types.
+     */
+    public static BytesRef stringToIP(BytesRef field) {
```
[the remark this answers is not in the document]
Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java:201]: > Nit: s/field/value (here and below), or something else, as it doesn't have to be a field, can be any value. But if we want these to be used mostly for cases where the input is field value, we might want to log a comment.

I'll try to give them more consistent and clear names. If it is a primitive/number, I'll call it number. If it is an instance of a class, I'll call it field. If there is something obviously better and meaningful, I'll use it in stead of a generic name.

[the remark this answers is not in the document]
Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java:201]: > Some of the functions here are just a renaming wrapper for functionality either in some utils or JDK, which I can see bring uniformity to conversion primitives in ESQL, but might seem like a bit of confusing duplication. So I was wondering if we don't want to keep this class focused on the converters [infrastructure] and move these utils somewhere (maybe QL, like to `StringUtils` and `NumericUtils` or some ESQL-specific ones).
> 
> Anyways, just mentioning, this can also stay as is.

Yeah, I was thinking about the same thing, it is a bit of confusing duplication for the conversions between primitives and those just call StringUtils and NumericUtils. As we are trying to make EsqlDataTypeConverter a central place for data type conversions within ESQL, I put all(or an anchor of those) of the conversions that I can find so far here. I was thinking of make a copy of the utility methods/functions used in QL into EsqlDataTypeConverter, but I haven't done it yet. There is a bigger effort of grand refactor #106679, perhaps it can be a subtask of it, so that it can be done in a more consistent way.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToLong.java (hunk 1 above)
Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToLong.java:94]: I'll make the change in this PR.
